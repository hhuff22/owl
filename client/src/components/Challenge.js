import React, { useState } from "react";
import axios from "axios";
import ItemRow from "./ItemRow";

export default function Challenge() {
  const [lowStock, setLowStock] = useState({
    data: "",
    loading: true,
  });
  const [restockCost, setRestockCost] = useState([]);

  const handleClickLowStock = async () => {
    const data = await axios.get("http://localhost:4567/low-stock");
    setLowStock({
      data: data.data,
      loading: false,
    });
  };

  const handleClickReorderCost = async () => {
    const data = await axios.post(
      "http://localhost:4567/restock-cost",
      lowStock.data
    );
    setRestockCost(data.data);
  };

  return (
    <>
      <table>
        <thead>
          <tr>
            <td>SKU</td>
            <td>Item Name</td>
            <td>Amount in Stock</td>
            <td>Capacity</td>
            <td>Order Amount</td>
          </tr>
        </thead>
        <tbody>
          {lowStock.loading
            ? ""
            : lowStock.data.map((item) => (
                <ItemRow
                  data-testid={"item-row"}
                  key={item.id}
                  item={item}
                  lowStock={lowStock}
                  setLowStock={setLowStock}
                />
              ))}
        </tbody>
      </table>
      <div class="totalCost" data-testid="total-cost">
        Total Cost: ${Math.round((restockCost + Number.EPSILON) * 100) / 100}
      </div>
      <button onClick={handleClickLowStock} data-testid="low-stock-button">Get Low-Stock Items</button>
      <button onClick={handleClickReorderCost} data-testid="reorder-button">Determine Re-Order Cost</button>
    </>
  );
}
