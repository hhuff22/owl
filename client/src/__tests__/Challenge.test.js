import React from 'react';
import Challenge from "../components/Challenge";
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import * as axios from "axios";
import '@testing-library/jest-dom/extend-expect';

jest.mock("axios");

const dummyGetResponseData = [{
    id: 0,
    name: "dummy item name",
    cost: 5
}];
const dummyPostResponseData = 4;

describe("Challenge", () => {
    test("should render correctly", () => {
        const { asFragment } = render(<Challenge />);
        expect(asFragment()).toMatchSnapshot();
    })

    test("should load low stock data", async () => {
        axios.get.mockImplementation(() => {
            return Promise.resolve({data:dummyGetResponseData});
        })
        const {getByTestId} = render(<Challenge />);
        const lowStockButton = screen.getByTestId("low-stock-button");
        fireEvent.click(lowStockButton, {target: {}});
        await waitFor(() => {
            const rows = getByTestId("item-row-0");
            expect(rows).toBeTruthy();
        })
        expect(lowStockButton).toBeTruthy();
    })

    test("should load reorder request data", async () => {
        axios.get.mockImplementation(() => {
            return Promise.resolve({data:dummyGetResponseData});
        })
        axios.post.mockImplementation(() => {
            return Promise.resolve({data:dummyPostResponseData});
        })
        const {getByTestId} = render(<Challenge />);
        const lowStockButton = screen.getByTestId("low-stock-button");
        fireEvent.click(lowStockButton, {target: {}});
        const reorderButton = screen.getByTestId("reorder-button");
        fireEvent.click(reorderButton, {target: {}});
        await waitFor(() => {
            const totalCostLabel = getByTestId("total-cost");
            expect(totalCostLabel).toHaveTextContent("Total Cost: $4");
        })
        expect(reorderButton).toBeTruthy();
    })
})