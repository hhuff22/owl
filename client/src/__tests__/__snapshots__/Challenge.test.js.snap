// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Challenge should render correctly 1`] = `
<DocumentFragment>
  <table>
    <thead>
      <tr>
        <td>
          SKU
        </td>
        <td>
          Item Name
        </td>
        <td>
          Amount in Stock
        </td>
        <td>
          Capacity
        </td>
        <td>
          Order Amount
        </td>
      </tr>
    </thead>
    <tbody />
  </table>
  <div
    class="totalCost"
    data-testid="total-cost"
  >
    Total Cost: $0
  </div>
  <button
    data-testid="low-stock-button"
  >
    Get Low-Stock Items
  </button>
  <button
    data-testid="reorder-button"
  >
    Determine Re-Order Cost
  </button>
</DocumentFragment>
`;
